src/ucs/datastruct/ptr_array.h

```diff
@@ -120,13 +150,229 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
                     (UCS_V_INITIALIZED(_var), 0) : \
                     !__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index])))
 
+/**
+ * Check if element is free, and calculate how many free elements are ahead.
+ *
+ * @param [in]     _elem  An element in the ptr array.
+ * @param [in/out] iter   Pointer to the foreach iterator, updated to next_idx.
+ *
+ * @return 1 if the element is free and 0 if it's occupied.
+ */
+static inline int
+__ucs_ptr_array_is_free_and_get_step_size(ucs_ptr_array_elem_t _elem,
+                                          uint32_t *_size_elem)
+{
+    int _is_free = ((uintptr_t)(_elem) & UCS_PTR_ARRAY_FLAG_FREE);
+
+    if (_is_free) {
+       *_size_elem = (_elem >> UCS_PTR_ARRAY_SIZE_FREE_SHIFT);
+    } else {
+       *_size_elem = 1;
+    }
+    ucs_assert(*_size_elem > 0);
+
+    return _is_free;
+}
 
 /**
  * Iterate over all valid elements in the array.
+ *
+ * @param [out] _var        Pointer to current array element in the foreach.
+ * @param [out] _index      Index variable to use as iterator (unsigned).
+ * @param [in]  _ptr_array  Pointer to a ptr array.
  */
 #define ucs_ptr_array_for_each(_var, _index, _ptr_array) \
-    for (_index = 0; _index < (_ptr_array)->size; ++_index) \
-         if (!__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index]))) \
+    (_index) = 0; \
+    for (uint32_t __foreach_step_size = 1; \
+            ((_index) < (_ptr_array)->size); (_index) += __foreach_step_size) \
+       if ((ucs_likely(!__ucs_ptr_array_is_free_and_get_step_size( \
```

Comment:
Need more whitespaces before `if` to ensure 4 spaces following `for`

Reply:
Fixed.